### PR TITLE
Allow nextgenrepl to real-time replicate reaps

### PR DESCRIPTION
**Pull request by martinsumner**
This is to address the issue of reaping across sync'd clusters.  Without this feature it is necessary to disable full-sync whilst independently replicating on each cluster.

Now if reaping via riak_kv_reaper the reap will be replicated assuming the `riak_kv.repl_reap` flag has been enabled.  At the receiving cluster the reap will not be replicated any further.

There are some API changes to support this.  The `find_tombs` aae_fold will now return Keys/Clocks and not Keys/DeleteHash.  The ReapReference for riak_kv_repaer will now expect a clock (version vector) not a DeleteHash, and will also now expect an additional boolean to indicate if this repl is a replication candidate (it will be false for all pushed reaps).

The object encoding for nextgenrepl now has a flag to indicate a reap, with a special encoding for reap references.

### CODE REVIEW

File: src/riak_object.erl [hunk 1]
```diff
@@ -1189,23 +1189,25 @@ binary_version(<<131,_/binary>>) -> v0;
 binary_version(<<?MAGIC:8/integer, 1:8/integer, _/binary>>) -> v1.
 
 %% @doc Encode for nextgen_repl
--spec nextgenrepl_encode(repl_v1, riak_object(), boolean()) -> binary().
+-spec nextgenrepl_encode(
+    repl_v1,
+    riak_object()|
```
Comment (ThomasArts) [inline src/riak_object.erl]: Look at type here... object or reap tuple. not a binary

File: src/riak_client.erl [hunk 2]
```diff
@@ -223,10 +224,11 @@ fetch(QueueName, {?MODULE, [Node, _ClientId]}) ->
 -spec push(riak_object:riak_object()|binary(),
```
Comment (ThomasArts) [inline src/riak_client.erl:224]: This type seems wrong... why does dialyzer not catch that.
`RObjMaybeBin` is either a riak_object or a binary.... For a binary representing a {reap,...} I would expect `riak_object:is_robject(....)` to return `false`.
So for those binaries you call:
```
riak_object:nextgenrepl_decode(RObjMaybeBin)
```
But in riak_object.erl line 1194 you gladly accept a tuple and not a binary.

File: src/riak_client.erl [hunk 3]
```diff
@@ -579,26 +600,30 @@ consistent_delete(Bucket, Key, Options, _Timeout, {?MODULE, [Node, _ClientId]})
...
             false
     end.
 
--spec reap(riak_object:bucket(), riak_object:key(), pos_integer(),
-                                                riak_client()) -> boolean().
-reap(Bucket, Key, DeleteHash, {?MODULE, [Node, _ClientId]}) ->
+-spec reap(
+    riak_object:bucket(), riak_object:key(), vclock:vclock(), riak_client())
+        -> boolean().
+reap(Bucket, Key, TombClock, {?MODULE, [Node, _ClientId]}) ->
```
Comment (ThomasArts) [inline src/riak_client.erl:616]: Above you replaced `{?MODULE, [Node, _ClientId]}` by RiakClient... probably you want that here as well and then access the Node via API function? Not sure, but it is kind of breaking abstraction boundary

File: src/riak_kv_replrtq_snk.erl [hunk 4]
```diff
@@ -721,11 +721,11 @@ repl_fetcher(WorkItem) ->
                             {tomb, FetchSplit, PushSplit, ModSplit});
             {ok, RObj} ->
                 SWFetched = os:timestamp(),
-                {ok, LMD} = riak_client:push(RObj, false, [], LocalClient),
                 SWPushed = os:timestamp(),
-                ModSplit = timer:now_diff(SWPushed, LMD),
                 FetchSplit = timer:now_diff(SWFetched, SW),
                 PushSplit = timer:now_diff(SWPushed, SWFetched),
+                {ok, LMD} = riak_client:push(RObj, false, [], LocalClient),
```
Comment (ThomasArts) [inline src/riak_kv_replrtq_snk.erl]: I guess this is right, but you now take the timewhen you start the push and before when the push replied.

File: src/riak_object.erl (hunk 1 above)
Comment (martinsumner) [reply src/riak_object.erl]: The output of the function is a binary() though, or have I missed your point?  Formatting of this spec is quite confusing perhaps, I'll address this.

**Review by ThomasArts — COMMENTED** [review]
Please have a look at the type conflict mentioned in comment
It might be me overlooking something, but worth a second pair of eyes

File: src/riak_client.erl (hunk 2 above)
Comment (ThomasArts) [reply src/riak_client.erl:224]: sorry, you call decode, not encode

**Review by ThomasArts — APPROVED** [review]

**Comment by martinsumner** [comment]
https://github.com/nhs-riak/riak_test/pull/4

**Comment by martinsumner** [comment]
https://github.com/nhs-riak/riak-erlang-client/pull/1

**Comment by martinsumner** [comment]
https://github.com/nhs-riak/riak-erlang-http-client/pull/1

**Review by stephenwood24-nhs — APPROVED** [review]

**Pull request merged**